pkg/lifecycle/lifecycle_provider.go

```diff
@@ -0,0 +1,78 @@
+package lifecycle
+
+import (
+	"sync/atomic"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	ConfigName = "lifecycle-image"
+	ConfigKey  = "image"
+)
+
+type RegistryClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, string, error)
+}
+
+type Provider struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+	lifecycleImage atomic.Value
+	callbacks      []func()
+	err            error
+}
+
+func NewProvider(lifecycleImageRef string, client RegistryClient, keychain authn.Keychain) (*Provider, error) {
+	img, _, err := client.Fetch(keychain, lifecycleImageRef)
+	if err != nil {
+		return &Provider{}, errors.Wrap(err, "failed to get lifecycle image")
+	}
+	p := &Provider{
+		RegistryClient: client,
+		Keychain:       keychain,
+	}
+	p.lifecycleImage.Store(img)
+	return p, nil
+}
+
+func (l *Provider) UpdateImage(cm *corev1.ConfigMap) {
```

Comment:
I'm trying to understand:

There's a **default lifecycle** that comes through flag/environment variables.
Now a user can specify **custom lifecycle** through configMap.
What happens when the controller is restarted? Which **lifecycle** wins?

Reply:
If the controller is restarted whatever the configmap value is would be mounted to the container.

I agree it's a little strange. For context, the only reason I added the seeding of the lifecycle at startup was because the kpack controller errors because the lifecycle provider hasn't updated before the builders reconcile and it's a potentially alarming error message ("lifecycle has not been loaded"). It eventually resolves it's just the log line.

Reply:
I noticed this option
```
configMapWatcher.WatchWithDefault(cm corev1.ConfigMap, o ...Observer)
```
vs what we have now
```
configMapWatcher.Watch(name string, o ...Observer)
```
 and was wondering if `WatchWithDefault` could be used for the lifecycle image passed through the flag/env   variables... pondering if it would be helpful

Reply:
It is important to note that the "default lifecycle" will be [mounted from the contents of the config map](https://github.com/pivotal/kpack/blob/46fe4fc3a99696d451a9e2fef2902451dd120108/config/controller.yaml#L110-L114).

Reply:
It might make sense for the provider to not execute callbacks when there is no. change to the image.

Reply:
@sukhil-suresh 
`configMapWatcher.WatchWithDefault` provides a nice initial default resource for the watcher, but it processes the same way as `configMapWatcher.Watch` meaning that we would still not guarantee the `configMapWatcher` runs an update before the `clusterbuilder` controller and we still get the initial "lifecycle is not loaded" error. ultimately `configMapWatcher.WatchWithDefault` works the same as `configMapWatcher.Watch` if the expected configmap is already there.